service/src/main/java/org/broadinstitute/listener/relay/inspectors/TokenChecker.java

```diff
@@ -16,43 +14,58 @@
 
 @Component
 public class TokenChecker {
-  private final HttpClient httpClient = HttpClient.newHttpClient();
-  private final String GOOGLE_OAUTH_SERVER =
-      "https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=";
   private final Logger logger = LoggerFactory.getLogger(SamResourceClient.class);
+  private final GoogleTokenInfoClient googleTokenInfoClient;
 
-  public OauthInfo getOauthInfo(String token) throws IOException, InterruptedException {
-    var jwtExpiration = tryDecodeAsB2cToken(token);
+  public TokenChecker(GoogleTokenInfoClient googleTokenInfoClient) {
+    this.googleTokenInfoClient = googleTokenInfoClient;
+  }
 
-    var now = Instant.now();
+  public OauthInfo getOauthInfo(String token) throws IOException, InterruptedException {
+    return getOauthInfoWithAnchorTimestamp(token, Instant.now());
+  }
 
-    if (jwtExpiration.isPresent()) {
-      if (jwtExpiration.get().isAfter(now)) return new OauthInfo(jwtExpiration, "");
-      else return new OauthInfo(Optional.empty(), "JWT expired");
+  public OauthInfo getOauthInfoWithAnchorTimestamp(String token, Instant anchor)
+      throws IOException, InterruptedException {
+    var jwt = tryDecodeAsB2CToken(token);
+    if (jwt.isPresent()) {
+      var jwtExpiration = jwt.get().getExpiresAt().toInstant();
+      if (jwtExpiration.isAfter(anchor)) {
+        var stringClaims = new HashMap<String, String>();
+        jwt.get().getClaims().forEach((k, v) -> stringClaims.put(k, v.asString()));
+        return new OauthInfo(Optional.of(jwtExpiration), "", stringClaims);
+      } else {
+        return new OauthInfo(Optional.empty(), "JWT expired", Map.of());
+      }
     } else {
-      var request = HttpRequest.newBuilder().uri(URI.create(GOOGLE_OAUTH_SERVER + token)).build();
-
-      var oauthInfoResponse = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
-      var decoded = new Gson().fromJson(oauthInfoResponse.body(), GoogleOauthInfoResponse.class);
+      var googleTokenInfo = googleTokenInfoClient.getTokenInfo(token);
+      if (googleTokenInfo.expires_in > 0) {
+        return new OauthInfo(
+            Optional.of(anchor.plusSeconds(googleTokenInfo.expires_in)),
+            googleTokenInfo.error,
+            claimsFromGoogleInfo(googleTokenInfo));
+      } else {
+        return new OauthInfo(Optional.empty(), googleTokenInfo.error, Map.of());
+      }
+    }
+  }
 
-      if (decoded.expires_in > 0)
-        return new OauthInfo(Optional.of(now.plusSeconds(decoded.expires_in)), decoded.error);
-      else return new OauthInfo(Optional.empty(), decoded.error);
+  private Map<String, String> claimsFromGoogleInfo(GoogleOauthInfoResponse response) {
+    var claims = new HashMap<>(Map.of("sub", response.user_id));
```

Comment:
I am unsure if this is semantically correct; from what I can tell, the google token info endpoint does not give back a `sub` field, but instead a `user_id` field. Happy to break this out into a separate field if need be.

Reply:
Good point -- I think this is correct to treat `user_id` from google as `sub`. We do this in [apache as well](https://github.com/broadinstitute/terra-helmfile/blob/master/charts/httpd-proxy/templates/_proxy_configmap.yaml#L158).